
####### Analyze output of HOMER annotatePeaks.pl to visualize genomic features overlapped by K36me1/2/3 peaks
#### Perform analysis for all peaks as well as top 10% peaks



rm(list=ls())

##### initialize general packages

library(rtracklayer)
library(GenomicFeatures)
library(RColorBrewer)
library(matrixStats)
library(grid)
library(gridBase)
library(gridExtra)
library(tsTools)
library(BiocParallel)
library(GenomicAlignments)
library(LSD)
library(csaw)
library(pheatmap)
library(Vennerable)
library(LSD)
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
library(IRanges)
library(ShortRead)
library(ComplexHeatmap)
library(circlize)
library(dendextend)
library(genefilter)
library(tidyverse)
library(zoo)
library(matrixStats)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(tsTools)
library(Vennerable)
library(HelpersforChIPSeq)

source("./functions/functions.R")
source("./functions/chromoMap_custom.R")


########
sample_peakfiles <- list.files(path="./Fly/Further_Analysis/BigWigs_Peaks/",pattern=".bed$",full.names = T)
library(ChIPseeker)


### read in peaks and generate files with top10% peaks -

i <- 1

for (i in seq_along(sample_peakfiles)){
  
  ## trim filenames accordingly here 
  name <- gsub(".bed","",gsub(".*2023//","",sample_peakfiles[i]))
  bedpeak <- read.delim(sample_peakfiles[i],header=F)
  
  ## select top 10% peaks based on peak strength
  bedpeak_top10 <- bedpeak %>% filter(V4 >= quantile(bedpeak$V4,probs=0.9)[[1]])
  write_delim(bedpeak_top10,paste0("./Fly/Further_Analysis/BigWigs_Peaks/",name,"_top10.bed"),col_names = F,quote = "none",delim = "\t")
  
}

########## RUN HOMER ANNOTATION OF PEAKS 

sample_peakfiles <- list.files(path="./Fly/Further_Analysis/BigWigs_Peaks/", pattern="peakAnno.txt$", full.names = T)
sample_peakfiles <- as.list(sample_peakfiles)
names(sample_peakfiles) <- gsub("_peakAnno.txt","",list.files(path="./Fly/Further_Analysis/BigWigs_Peaks/", pattern="peakAnno.txt$"))
library(data.table)
#peakAnnoList_Homer <- lapply(sample_peakfiles,fread, select=c("Annotation"))
peakAnnoList_Homer <- rbindlist(sapply(sample_peakfiles, fread,simplify = FALSE, select="Annotation"), idcol = 'PeakType')

library(forcats)

#####simplify annotations generated by HOMER by grouping together 

peakAnnoList_Homer$Annotation_Grouped <-fct_collapse(peakAnnoList_Homer$Annotation,
                                                     intron=unique(peakAnnoList_Homer$Annotation[grep("intron",peakAnnoList_Homer$Annotation)]),
                                                     exon=unique(peakAnnoList_Homer$Annotation[grep("exon",peakAnnoList_Homer$Annotation)]),
                                                     `3_UTR`=unique(peakAnnoList_Homer$Annotation[grep("3'",peakAnnoList_Homer$Annotation)]),
                                                     `5_UTR`=unique(peakAnnoList_Homer$Annotation[grep("5'",peakAnnoList_Homer$Annotation)]),
                                                     TTS=unique(peakAnnoList_Homer$Annotation[grep("TTS",peakAnnoList_Homer$Annotation)]),
                                                     Intergenic=unique(peakAnnoList_Homer$Annotation[grep("Intergenic",peakAnnoList_Homer$Annotation)]),
                                                     Promoter_TSS=unique(peakAnnoList_Homer$Annotation[grep("promoter-TSS",peakAnnoList_Homer$Annotation)]))
peakAnnoList_Homer <- peakAnnoList_Homer[!is.na(peakAnnoList_Homer$Annotation)]
peakAnnoList_Homer_counts <- peakAnnoList_Homer %>% group_by(PeakType) %>% count(Annotation_Grouped)

#### visualize as barplots stacked

theme_bars1 <- theme_grey(base_size = 10, base_family = "") %+replace%
  theme(
    axis.title = element_text(size = rel(1.5)),
    axis.text = element_text(size = rel(1)),
    axis.ticks.x = element_line(colour = "black", size= rel(0.7)),
    axis.ticks.y = element_blank(),
    panel.background = element_rect(fill = "transparent", colour = NA),
    plot.background = element_rect(fill = "transparent", colour = NA),
    strip.background = element_blank(),
    strip.text =  element_text(size = rel(1)))
p <- ggplot(peakAnnoList_Homer_counts, aes(x="",y=n, fill=Annotation_Grouped)) + geom_bar(stat="identity",position="fill") + facet_grid(~PeakType, scales = "free", space="free")+
  scale_x_discrete(expand=c(0,0.5)) +
  coord_cartesian(expand=F)
p <- p + facet_wrap(~PeakType, scales ="free",strip.position = "left", nrow=6, ncol=1)
p <- p + labs(title = "Peak distribution",x="",y="")
chr.col <- function(feature) {
  return(switch(feature,
                "3_UTR" = "#1F78B4",
                "5_UTR" = "#f5e618",
                "exon" = "#E31A1C",
                "Intergenic" = "#FF7F00",
                "intron" = "#B15928",
                "Promoter_TSS" = "#6A3D9A",
                "TTS" = "#33A02C"))
}
cols <- sapply(unique(peakAnnoList_Homer_counts$Annotation_Grouped), chr.col)
p1 <- p + scale_fill_manual(values=cols) + theme_bars1
p1 <- p1 + coord_flip()
p1 <- p1 + theme(strip.text.y.left = element_text(size=rel(1),angle=0))
pdf("./Fly/Further_Analysis/BigWigs_Peaks/PeakAnnoGenome_HOMER.pdf")
p1
dev.off()

